source/common/io/io_uring.h

```diff
@@ -0,0 +1,107 @@
+#pragma once
+
+#include "envoy/common/pure.h"
+
+#include "source/common/network/address_impl.h"
+
+namespace Envoy {
+namespace Io {
+
+/**
+ * Callback invoked when iterating over entries in the completion queue.
+ * @param user_data is any data attached to an entry submitted to the submission
+ * queue.
+ * @param result is a return code of submitted system call.
+ */
+using CompletionCb = std::function<void(void* user_data, int32_t result)>;
+
+enum class IoUringResult { Ok, Busy, Failed };
+
+/**
+ * Abstract wrapper around `io_uring`.
+ */
+class IoUring {
+public:
+  virtual ~IoUring() = default;
+
+  /**
+   * Registers an eventfd file descriptor for the ring and returns it.
+   * It can be used for integration with event loops.
+   */
+  virtual os_fd_t registerEventfd() PURE;
+
+  /**
+   * Resets the eventfd file descriptor for the ring.
+   */
+  virtual void unregisterEventfd() PURE;
+
+  virtual bool isEventfdRegistered() const PURE;
+
+  /**
+   * Iterates over entries in the completion queue, calls the given callback for
+   * every entry and marks them consumed.
+   */
+  virtual void forEveryCompletion(CompletionCb completion_cb) PURE;
+
+  /**
+   * Prepares an accept system call and puts it into the submission queue.
+   * Returns IoUringResult::Failed in case the submission queue is full already
+   * and IoUringResult::Ok otherwise.
+   */
+  virtual IoUringResult prepareAccept(os_fd_t fd, struct sockaddr* remote_addr,
+                                      socklen_t* remote_addr_len, void* user_data) PURE;
```

Comment:
Particularly this interface hides the io_uring syscalls and makes unit testing of the new iohandle (not in this PR) easier a bit. Originally I used a custom type containing the network iohandle instead of *void, but in case of Block IO a different type might be needed. Probably this *void in the interface could be avoided with a template.

I haven't thought of the possibility to implement an alternative Dispatcher, because its interface implies the file readiness model (i.e. Dispatcher::createFileEvent()), whereas io_uring is all about completeness of system operations. CompletionCb is an analog of FileReadyCb, but the former isn't called for a known file descriptor. A system operation and its context (including the arguments) are parts of a completeness event.

Reply:
I think it's ok to use void* for now, we may revisit this later as our understanding of this module improves.

I'ld like to better understand how io_uring and dispatcher should interact.  I could see some threads being exclusively one or the other.  I'm not sure about how they would coexist in a thread while avoiding starvation.

Reply:
I see no problem of multiple io_urings coexisting in one dispatcher: all updates in their completion queues are reflected in their respective event notification file descriptors (`man eventfd`). So, different callback instances are called for different eventfds, only one registered callback per io_uring. The dispatcher is supposed to call these callbacks in the same manner as in the case of connection callbacks.

I have rewritten the tests to run the callbacks through Dispatcher::run().

Though I fail to come up with a simple enough API for Dispatcher to accommodate multiple io_urings if go this way. If we assume only one io_uring per Dispatcher then extending Dispatcher with something like `void Dispatcher::registerCompletionEvent(CompletionCb cb)` and `IoUring& Dispatcher::getIoUring()` would be enough, `IoUring::registerEventfd()` could be dropped, since all manipulations with eventfd would be incapsulated in Dispatcher.

For the case of multiple io_urings `Dispatcher::getIoUring()` and `Dispatcher::registerCompletionEvent()` should distinguish io_urings somehow. Maybe something like the following could work:

```c++
using IoUringId = uint32_t;

class DispatcherImpl {
public:
  ...
  IoUringId addIoUring(const IoUringFactory& factory);
  void registerCompletionEvent(IoUringId id, CompletionCb cb);
  IoUring& getIoUring(IoUringId id);
  ...

private:
  std::vector<IoUring> io_urings_;
};
```